// @lemoncode/fonk-required-by-field-validator/dist/@lemoncode/fonk-required-by-field-validator.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var fonk = require('@lemoncode/fonk');

function _typeof(obj) {
  "@babel/helpers - typeof";

  if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
    _typeof = function (obj) {
      return typeof obj;
    };
  } else {
    _typeof = function (obj) {
      return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;
    };
  }

  return _typeof(obj);
}

var VALIDATOR_TYPE = 'REQUIRED_BY_FIELD';
var defaultMessage = 'Please fill in this mandatory field.';
var setErrorMessage = function setErrorMessage(message) {
  return defaultMessage = message;
};

var isDefined = function isDefined(value) {
  return value !== void 0 && value !== null && value !== '';
};

var isDefinedCustomArgs = function isDefinedCustomArgs(customArgs) {
  return isDefined(customArgs) && (isDefined(customArgs.field) && isDefined(customArgs.value) || isDefined(customArgs.field) && isDefined(customArgs.condition));
};

var checkRequiredParameters = function checkRequiredParameters(customArgs, values) {
  if (!isDefinedCustomArgs(customArgs)) {
    throw "Must provide valid customArgs: { field: '', value: '' } or { field: '', condition: () => true } in ValidationSchema";
  }

  if (!isDefined(values) || _typeof(values) !== 'object') {
    throw "Must provide valid \"values\" object";
  }
};

var getValueByCustomArgs = function getValueByCustomArgs(values, _ref) {
  var field = _ref.field;
  return field.split('.').reduce(function (value, key) {
    return value ? value[key] : null;
  }, values);
};

var hasToBeRequired = function hasToBeRequired(values, customArgs) {
  var customArgFieldValue = getValueByCustomArgs(values, customArgs);
  return isDefined(customArgs.value) ? customArgFieldValue === customArgs.value : customArgs.condition(customArgFieldValue, values);
};

var validator = function validator(fieldValidatorArgs) {
  var value = fieldValidatorArgs.value,
      values = fieldValidatorArgs.values,
      _fieldValidatorArgs$m = fieldValidatorArgs.message,
      message = _fieldValidatorArgs$m === void 0 ? defaultMessage : _fieldValidatorArgs$m,
      customArgs = fieldValidatorArgs.customArgs;
  checkRequiredParameters(customArgs, values);
  var succeeded = true;

  if (hasToBeRequired(values, customArgs)) {
    var validationResult = fonk.Validators.required.validator({
      value: value,
      values: values,
      customArgs: customArgs
    });
    succeeded = validationResult.succeeded;
  }

  return {
    succeeded: succeeded,
    message: succeeded ? '' : fonk.parseMessageWithCustomArgs(message || defaultMessage, customArgs),
    type: VALIDATOR_TYPE
  };
};

var validator$1 = /*#__PURE__*/Object.freeze({
  __proto__: null,
  setErrorMessage: setErrorMessage,
  validator: validator
});

exports.requiredByField = validator$1;
